Build tool fixes for Musca support

deepikabhavnani opened the pull request:
### Description

Typos and minor fixes required to support Cortex-M33 devices. Support of M33 was not tested since we didn;t had any target. This PR will be tested internally with MUSCA.

### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Feature
    [ ] Breaking change

Commit:
Fix build and typo issues with M33 build

Commit:
-mcpu option not needed for v8M devices

-mcpu option if set for v8M CPU;s it will add DSP feature as default
which is optional. Hence setting just the architecture for Cortex-M23
and Cortex-M33

cmonr commented on tools/toolchains/arm.py:
```diff
@@ -393,14 +393,11 @@ def __init__(self, target, *args, **kwargs):
-            self.flags['common'].append("-mcpu=%s" % target.core.lower())
-            self.flags['ld'].append("--cpu=%s" % target.core.lower())
-            self.SHEBANG += " -mcpu=%s" % target.core.lower()
+            if not target.core.startswith("Cortex-M23") and not target.core.startswith("Cortex-M33"):
```
Hmm, so I take it building for the NS side no longer needs to be differentiated?

deepikabhavnani replied:
Differentiation is still there, here we are just not setting ‘-mcpu’ parameter for any of v8M device. If we set cpu as cortex-m33 in flags, compiler enables dsp as well which is optional. Hence we just set ‘-march’

theotherjimmy replied:
Please collapse the 
```
else:
    if: ...
```
into 
```
elif:
```

cmonr commented on tools/toolchains/gcc.py:
```diff
@@ -256,8 +256,6 @@ def link(self, output, objects, libraries, lib_dirs, mem_map):
         # Exec command
         self.notify.cc_verbose("Link: %s" % ' '.join(cmd))
         self.default_cmd(cmd)
-        if self.target.core == "Cortex-M23" or self.target.core == "Cortex-M33":
```
Why is this output no longer needed? (Commit only mentioned something about a build issue)

deepikabhavnani replied:
Code for creation of ‘secure_file’ was initially in this section and later moved to other part. Variable ‘secure_file’ is not available here and hence throws error

cmonr reviewed (commented):
LGTM, just some questions

Commit:
Arch and CPU options for linker and Clang are different for Cortex-M33

As per the link below, options for clang and armlink are diferrent for Cortex-M33
armlink --cpu 8-M.Main --import-cmse-lib-out importlib_v1.o
armclang -march=armv8-m.main -mcmse

http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0773h/pge1452794854109.html

deepikabhavnani commented:
Fix in https://github.com/ARMmbed/mbed-os/pull/7792/commits/63664520c028f7c704ae5f4f4e86af6e2784e36a : Options for clang and armlink are diferrent for Cortex-M33
```
armlink --cpu 8-M.Main --import-cmse-lib-out importlib_v1.o
armclang -march=armv8-m.main -mcmse

```
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0773h/pge1452794854109.html

cc @gaborkertesz

deepikabhavnani commented:
@theotherjimmy - We have few issues while building ASM files. All the C flags are added to asm build, as a result .o generated is not correct.

`Compile: C:\Program Files\ARMCompiler6.10\bin\armasm --cpu=8-M.Main --cpreproc --**cpreproc_opts=-c**,--target=arm-arm-none-eabi,-mthumb,-Os,-Wno-armcc-pragma-push-pop,-Wno-armcc-pragma-anon-unions,-DMULADDC_CANNOT_USE_R7,-fdata-sections,-fno-exceptions,-MMD,-D_LIBCPP_EXTERN_TEMPLATE(...)=,-D__STARTUP_COPY_MULTIPLE,-D__CORTEX_M33,-DMBED_FAULT_HANDLER_DISABLED,-D__MBED_CMSIS_RTOS_CM,-D__STARTUP_CLEAR_BSS,-D__CMSIS_RTOS,-DARM_MATH_ARMV8MML,-I.,-I.\mbed-os,-I.\mbed-os\cmsis,-I.\mbed-os\cmsis\TARGET_CORTEX_M,-I.\mbed-os\drivers,-I.\mbed-os\events,-I.\mbed-os\events\equeue,-I.\mbed-os\features,-I.\mbed-os\features\cellular,-I.\mbed-os\features\cellular\easy_cellular, 
`

```
[DEBUG] Output: armclang.exe: warning: .\mbed-os\targets\TARGET_ARM_SSG\TARGET_MUSCA_A1\TOOLCHAIN_ARMC6\startup_cmsdk_musca.s: 'assembler' input unused [-Wunused-command-line-argument]
[DEBUG] Output: armclang.exe: warning: argument unused during compilation: '-c' [-Wunused-command-line-argument]
[DEBUG] Output: armclang.exe: warning: argument unused during compilation: '-mthumb' [-Wunused-command-line-argument]
[DEBUG] Output: armclang.exe: warning: argument unused during compilation: '-Wno-armcc-pragma-push-pop' [-Wunused-command-line-argument]
[DEBUG] Output: armclang.exe: warning: argument unused during compilation: '-Wno-armcc-pragma-anon-unions' [-Wunused-command-line-argument]
[DEBUG] Output: armclang.exe: warning: argument unused during compilation: '-D MULADDC_CANNOT_USE_R7' [-Wunused-command-line-argument]
[DEBUG] Output: armclang.exe: warning: argument unused during compilation: '-fdata-sections' [-Wunused-command-line-argument]
[DEBUG] Output: armclang.exe: warning: argument unused during compilation: '-fno-exceptions' [-Wunused-command-line-argument]
```

Commit:
Remove deprecated linker options

Resolved below warnings
Warning: L3912W: Option 'legacyalign' is deprecated.
Warning: L3912W: Option 'no_strict_wchar_size' is deprecated.
Warning: L3912W: Option 'no_strict_enum_size' is deprecated.

deepikabhavnani commented:
> We have few issues while building ASM files. All the C flags are added to asm build, as a result .o generated is not correct.

Issue relate to ASM file build was related to assembly file extension ".s", and is resolved by updating it to be caps .S

deepikabhavnani commented:
@theotherjimmy  @gaborkertesz  - Please review

Commit:
Disabling DSP and FPU for Cortex-M33

Change shared by @Gabor Kertesz

Hard-fault on Musca was because of enabled floating point instructions,
disabling DSP and FPU.

gaborkertesz commented:
Basic hello world example has been tested with armclang and gcc on Musca and working as expected as a secure only application.

deepikabhavnani commented:
@theotherjimmy  @cmonr - Please review

theotherjimmy commented on tools/toolchains/arm.py:
```diff
@@ -414,10 +412,14 @@ def __init__(self, target, *args, **kwargs):
+            self.flags['c'].append("-mcpu=cortex-m33+nodsp")
+            self.flags['c'].append("-mfpu=none")
+            self.flags['cxx'].append("-mcpu=cortex-m33+nodsp")
+            self.flags['cxx'].append("-mfpu=none")
```
Can these be added to common?

deepikabhavnani replied:
Done 👍

theotherjimmy commented on tools/toolchains/arm.py:
```diff
@@ -414,10 +412,14 @@ def __init__(self, target, *args, **kwargs):
         if target.core == "Cortex-M23" or target.core == "Cortex-M33":
-            self.flags['common'].append("-mcmse")
+            self.flags['cxx'].append("-mcmse")
+            self.flags['c'].append("-mcmse")
```
Could this be added to common?

deepikabhavnani replied:
This flag gives warning with linker and armasm.

0xc0170 commented:
I got one question - removing those legacy flags from linker command - the warning is for ARMC6 only (can't be seen in ARMC5) ? This is safe to remove, no functionality change?

deepikabhavnani commented:
> I got one question - removing those legacy flags from linker command 

Legacy linker flags are not present for ARM 5 compiler,
https://github.com/ARMmbed/mbed-os/pull/7792/files#diff-84bdeb5b184eec3910a5d6a23df9ceccL35

I don;t see those flags even in history of ARM 5 compiler, @theotherjimmy - any idea why those flags were added for ARMC6.

>  This is safe to remove

Should be, as they have no impact. Deprecated as per this link http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0381b/CHDHCGGF.html

deepikabhavnani commented:
@theotherjimmy - I also see below warning in detail build dumps, not sure what is causing this.

`[DEBUG] Errors: warning: 'armv4t' is unsupported in this version of the product`

deepikabhavnani commented:
Can we start the build for this PR?

cmonr commented:
Does https://github.com/ARMmbed/mbed-os/pull/7792#issuecomment-415430907 not need to be addressed?

deepikabhavnani commented:
> Does #7792 (comment) not need to be addressed?

Its warning, can be fixed later as well. Musca is dependent on this fix for now. Added issue to handle it later https://github.com/ARMmbed/mbed-os/issues/7900

cmonr commented:
/morph build
